### PR TITLE
[autograd] fix allow_unused checking for C++ API

### LINKED ISSUE #34029: A minor bug in unused grad_input check of run_backward()
## 🐛 Bug

A pretty obvious bug that `num_inputs` is always 0.

https://github.com/pytorch/pytorch/blob/877c96cddfebee00385307f9e1b1f3b4ec72bfdc/torch/csrc/autograd/autograd.cpp#L103-L113

cc @ezyang @SsnL @albanD @zou3519 @gqchen

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34035 [autograd] fix allow_unused checking for C++ API**

Bug for the conditon check in https://github.com/pytorch/pytorch/pull/24342, realized we don't have tests in either
python or cpp to catch this, so added testes for both python and cpp.

Thanks @hczhu on capturing it!

Differential Revision: [D20198837](https://our.internmc.facebook.com/intern/diff/D20198837)

### CODE REVIEW

**Review by hczhu — APPROVED** [review]

File: torch/csrc/autograd/autograd.cpp [hunk 1]
```diff
@@ -100,7 +100,7 @@ variable_list run_backward(
   variable_list grad_inputs = Engine::get_default_engine().execute(
       roots, grad_outputs, keep_graph, create_graph, output_edges);
   // check if grad_inputs contains None or not base on the allow_unused flag
-  if (inputs.empty()) {
+  if (!inputs.empty()) {
```
Comment (hczhu) [inline torch/csrc/autograd/autograd.cpp]: nit:
skip the check if `allow_unused` is true.
```
if (!inputs.empty() && !allow_unused) {

```

**Comment by albanD** [comment]
@wanchaol this is BC-breaking change right? This will raise errors for user code that used to run.

**Comment by wanchaol** [comment]
> @wanchaol this is BC-breaking change right? This will raise errors for user code that used to run.

@albanD hmm yeah, but this is only BC-breaking for the autograd C++ API (not the one from python as they are not sharing the logic, and C++ API did it wrong for this check). Also I would say if user are using it that way it should be wrong usage at first (i.e. the same thing will throw error when using python).

**Review by albanD — APPROVED** [review]
Same nit as the one above.
I got confused by the fact that you added a test for the python code. But it was already working. Sounds good.

**Pull request closed without merging**